### PR TITLE
ENH: Add index/rindex ufuncs for unicode and bytes dtypes

### PR DESCRIPTION
**Pull request by lysnikolaou**

### CODE REVIEW

**Review by ngoldbaum — COMMENTED** [review]
This all looks good to me! Once #25347 is merged I will push commits to this PR adding loops for stringdtype and then this can be merged.

**Comment by ngoldbaum** [comment]
Looking at this again, I think it would be better to refactor the loops for `find` and `rfind` so we can reuse that code, since this is almost the same. Take a look at my last push that does this for the stringdtype `find` and `rfind` loops. Also I rebased on main to get stringdtype in this branch, so you'll need to pull from github on your local branch. I probably should have merged but I wasn't thinking, sorry!

**Comment by ngoldbaum** [comment]
Also maybe some unicode characters in the tests too? The stringdtype tests have some so not a biggie.

File: numpy/_core/code_generators/generate_umath.py
Comment (ngoldbaum) [inline numpy/_core/code_generators/generate_umath.py:1210]: good catch!

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (ngoldbaum) [inline numpy/_core/code_generators/ufunc_docstrings.py:4625]: Commenting to remind myself to update all of these to mention stringdtype as well. No need to fix this yourself.

**Review by ngoldbaum — APPROVED** [review]
LGTM. @mhvk would you like to take a look at this as well?

File: numpy/_core/src/umath/string_ufuncs.cpp
Comment (mhvk) [inline numpy/_core/src/umath/string_ufuncs.cpp:358]: Are we allowed to raise inside the loop? Don't you need `npy_gil_error` here too?

**Review by mhvk — COMMENTED** [review]
A comment in-line.

Two general comments:
1. About the implementation, I think using a template implies these pieces of code get compiled to two different units, i.e., essentially doubling the compiled code, while if we passed in `raise_error` through `aux_data` it would not (for a minimal decrease in speed, since `string_find` itself is much slower than checking a `bool`).
2. Do we really need this? It was there for `np.char`, possibly just because `str` has it. But do we care enough about failing fast? Or can we just use a python wrapper that checks whether any of the outputs is `-1`? Given that it is a simple addition, perhaps no worry, but I do wonder what the use case is for an array of strings...

**Comment by lysnikolaou** [comment]
> About the implementation, I think using a template implies these pieces of code get compiled to two different units, i.e., essentially doubling the compiled code, while if we passed in raise_error through aux_data it would not (for a minimal decrease in speed, since string_find itself is much slower than checking a bool).

That's true. I'll try this out as soon as your PR that does this for stringdtype gets merged.

> Do we really need this? It was there for np.char, possibly just because str has it. But do we care enough about failing fast? Or can we just use a python wrapper that checks whether any of the outputs is -1? Given that it is a simple addition, perhaps no worry, but I do wonder what the use case is for an array of strings...

That's a good point, although this matches the CPython API with a ufunc for very little effort, so I'm wondering if it'd hurt either way.

**Comment by ngoldbaum** [comment]
Marten do you mind if we merge this now so we get fast ufuncs for these operations in the NumPy 2.0 RC? We can do the refactor to use auxdata in a followup.

**Comment by mhvk** [comment]
I'd prefer to try to get this right in one go, which would mean merging #25796 first and then rebasing this one and use that framework. This of course partially is laziness, as doing this one first means quite a bit more work for me and I really should spend a bit more time on my astronomy job... (though admittedly the main work was finding out how to pass along method function pointers).

On the other hand, I see the need for getting things in. How about this, we ask @seberg if he's happy with the changes to the dtype API in #25796, so that that one can indeed be merged? And if he'd rather take more time, we put this one in first.

**Comment by lysnikolaou** [comment]
I adjusted this to the new `static_data` approach, so it's ready for another review.

**Comment by lysnikolaou** [comment]
I'm not sure what to make of the `benchmark` CI failure. I skimmed through the logs and didn't see something that might be related to this PR, but I might be wrong.

File: numpy/_core/src/umath/string_ufuncs.cpp
Comment (mhvk) [inline numpy/_core/src/umath/string_ufuncs.cpp:362]: I don't quite know whether `PyErr_Occurred()` is fast or slow, but in principle one could add an additional bit of data to `static_data` that tells whether this test is necessary. Alternatively, maybe that is still a good reason for expanding the template to `<ENCODING end, bool can_py_err>`?

On the other hand, even counting is probably slow compared to checking for the error; surely this is optimized quite well by python! So, definitely fine also to leave as is.

File: numpy/_core/tests/test_strings.py
Comment (mhvk) [inline numpy/_core/tests/test_strings.py:571]: I'd tend to replace `MAX` with `None` since that is the actual call signature.

**Review by mhvk — APPROVED** [review]
Looks great! Neither comment needs to be addressed (the first I think is actually better not to address , but I left it in for you to have a look), so I'll approve now

File: numpy/_core/src/umath/string_ufuncs.cpp
Comment (seberg) [reply numpy/_core/src/umath/string_ufuncs.cpp:362]: Came from an email:  Do you even hold the GIL to check (which is needed)?  Although I thought python debug would flush such an error out.

Comment (lysnikolaou) [reply numpy/_core/src/umath/string_ufuncs.cpp:362]: > Do you even hold the GIL to check (which is needed)?

Oh, that's a good point. Maybe it indeed makes sense to just have an additional bit of info in `static_data`.

Comment (seberg) [reply numpy/_core/src/umath/string_ufuncs.cpp:362]: Sorry, but that doesn't work: the reason why `static_data` is called `static` :).  

That is why I thought adding some `scratch_data` to the `context` is likely useful (but I didn't have a pressing enough need to actually add it yet, and I always tried to be minimal).  Which you can feel free to do:  A uint64 or two with the only guarantee that NumPy will NULL it before the first call.

Comment (seberg) [reply numpy/_core/src/umath/string_ufuncs.cpp:362]: In case it wasn't clear: The reason is that using that space wouldn't be thread-safe.

Comment (mhvk) [reply numpy/_core/src/umath/string_ufuncs.cpp:362]: It does work - the static data can tell whether you are in `index` (which can do `npy_gil` stuff) or in regular `find` (which cannot).

But I do think this points to the larger problem - can one check `npy_gil...` has been called in the function?

Comment (seberg) [reply numpy/_core/src/umath/string_ufuncs.cpp:362]: Ah, sorry, I thought you needed to keep track of whether an error occurred or not.  The nicest thing would be transition to new loops and just fail directly.

Comment (lysnikolaou) [reply numpy/_core/src/umath/string_ufuncs.cpp]: I ended up going with just returning `-2` from `index`/`rindex` as a sentinel for exception raised. Does that work?

File: numpy/_core/code_generators/ufunc_docstrings.py
Comment (ngoldbaum) [reply numpy/_core/code_generators/ufunc_docstrings.py]: I just pushed a commit to this branch to fix these docstrings.

**Comment by ngoldbaum** [comment]
I'll merge this in a little while assuming the tests pass after the last fixup commit for the docstrings I just pushed.

**Review by mhvk — APPROVED** [review]
I like the returning `-2` - that seems much cleaner! I think that captures everything, so approving. Thanks!

**Pull request merged**
